File: iosApp/iosApp/UI/GiphyViewModel.swift
//
//  GiphyViewModel.swift
//  iosApp
//
//  Created by USER on 2023/03/28.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import Combine
import shared

public class GiphyViewModel: ObservableObject {
    private let GIF_INITIAL_TITLE = "Giphy Random GIFs"
    private let GIF_SEARCH_PREFIX = "Search "
    private let GIF_SEARCH_POSTFIX = " ..."
    private let GIF_RANDOM_MAX_COUNT = 30
    
    private var giphySharedEngine = GiphySharedEngine()
    
    private var timer: Timer = Timer()
    
    private var searchOffset: Int32 = 0
    
    @Published
    var giphyTitle: String
    @Published
    var searchMode: SearchMode
    @Published
    var gifList: [GifUiModel] = []
    @Published
    var autoCompleteList: [GifAutoCompleteUiModel] = []
    @Published
    var autoCompleteListVisibility: ViewVisibility = .gone
    
    private var disposeBag = CFlowDisposebag()
    
    init() {
        self.giphyTitle = GIF_INITIAL_TITLE
        self.searchMode = SearchMode.Debouncing
    }
    
    func initViewModel() {
        getRandomGifList(maxCount: GIF_RANDOM_MAX_COUNT)
    }
    
    func updateTitle(query: String) {
        if (query.isEmpty) {
            giphyTitle = GIF_INITIAL_TITLE
        } else {
            giphyTitle = GIF_SEARCH_PREFIX + query + GIF_SEARCH_POSTFIX
        }
    }
    
    func updateSearchQuery(query: String) {
        timer.invalidate()
        clearPreviousSearchResult()
        searchOffset = 0

        if (query.isEmpty) {
            getRandomGifList(maxCount: GIF_RANDOM_MAX_COUNT)
        } else {
            getGifObjectsFromQuery(query: query)
            getAutoCompleteTerms(query: query)
        }
    }
    
    
    func loadMoreItem(searchOffset: Int32, query: String) {
        self.searchOffset = searchOffset
        updateSearchQuery(query: query)
    }
    
    func swapSearchMode() {
        if (searchMode == SearchMode.Debouncing) {
            searchMode = SearchMode.Throttling
        } else {
            searchMode = SearchMode.Debouncing
        }
    }
    
    func setAutoCompleteVisibility(visible: Bool) {
        if (visible) {
            autoCompleteListVisibility = .visible
        } else {
            autoCompleteListVisibility = .gone
        }
    }
    
    private func getRandomGifList(maxCount: Int) {
        var count = maxCount
        timer = Timer.scheduledTimer(withTimeInterval: 0.2, repeats: true) { timer in
            count = count - 1
            self.getRandomGif(timer: timer, count: count)
        }
    }
    
    private func getRandomGif(timer: Timer, count: Int) {
        (giphySharedEngine.getRandomGif() as! CFlow<GifEntity>)
            .watch { (gifEntity: GifEntity?) in
                if (count == 0) {
                    timer.invalidate()
                    return
                }
                if let entity = gifEntity {
                    self.gifList.append(GifUiModel(title: entity.title, url: entity.url, downsizedUrl: entity.downsizedUrl))
                }
            }.store(in: disposeBag)
    }
    
    private func getGifObjectsFromQuery(query: String) {
        (giphySharedEngine.getGifFromSearchQuery(query: query, offset: searchOffset) as! CFlow<NSArray>)
            .watch { (array: NSArray?) in
                if let gifEntityList = array as? [GifEntity] {
                    self.gifList.append(contentsOf: gifEntityList.map { entity in
                        return GifUiModel(title: entity.title , url: entity.url, downsizedUrl: entity.downsizedUrl)
                    })
                }
            }.store(in: disposeBag)
    }
    
    
    private func getAutoCompleteTerms(query: String) {
        (giphySharedEngine.getAutoCompleteTerms(query: query) as! CFlow<NSArray>)
            .watch { (array: NSArray?) in
                if let entityList = array as? [AutoCompleteEntity] {
                    self.setAutoCompleteVisibility(visible: true)
                    self.autoCompleteList = entityList.map { entity in
                        return GifAutoCompleteUiModel(name: entity.name)
                    }
                }
            }.store(in: disposeBag)
    }
    
    private func clearPreviousSearchResult() {
        gifList.removeAll()
        autoCompleteList.removeAll()
        setAutoCompleteVisibility(visible: false)
        
    }
    
    
    deinit {
        disposeBag.dispose()
    }
}

File: compose_desktop/src/jvmMain/kotlin/GiphyViewModel.kt
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import com.example.giphy_kmm.GiphySharedEngine
import com.example.giphy_kmm.data.SearchMode
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import model.GifAutoCompleteUiModel
import model.GifUiModel
import org.slf4j.event.Level
import java.util.logging.Logger

class GiphyViewModel {
    private val logger = Logger.getLogger(":::")
    private val giphyEngine = GiphySharedEngine()

    private val allApiJob: Job = Job()
    private val defaultDispatcher = Dispatchers.Default + allApiJob

    private var randomApiJob: Job? = null
    private var queryApiJob: Job? = null
    private var autoCompleteApiJob: Job? = null

    val giphyTitle = mutableStateOf(GIF_INITIAL_TITLE)

    private val _gifListStateFlow = MutableStateFlow<List<GifUiModel>>(listOf())
    val gifListStateFlow: StateFlow<List<GifUiModel>> get() = _gifListStateFlow

    val searchQuery = mutableStateOf("")
    val autoTermsList = mutableStateListOf<GifAutoCompleteUiModel>()
    val autoTermListShow = mutableStateOf(false)
    val searchMode = mutableStateOf(SearchMode.DEBOUNCING)

    private var searchOffset = 0

    fun initViewModel() {
        logger.log(java.util.logging.Level.ALL, "initViewModel")
        getRandomGifList(GIF_COUNT)
    }

    fun updateSearchQuery(query: String) {
        CoroutineScope(defaultDispatcher).launch {
            randomApiJob?.cancelAndJoin()
            queryApiJob?.cancelAndJoin()
            autoCompleteApiJob?.cancelAndJoin()

            _gifListStateFlow.emit(listOf())
            autoTermsList.clear()

            searchOffset = 0
            setAutoCompleteVisibility(false)

            searchQuery.value = query

            if (query.isEmpty()) {
                getRandomGifList(GIF_COUNT)
            } else {
                getGifObjectFromQuery(query)
                getAutoCompleteTerms(query)
            }

            updateTitle(query)
        }
    }

    fun swapSearchMode() {
        if (searchMode.value == SearchMode.DEBOUNCING) {
            searchMode.value = SearchMode.THROTTLING
        } else {
            searchMode.value = SearchMode.DEBOUNCING
        }
    }

    fun setAutoCompleteVisibility(visible: Boolean) {
        autoTermListShow.value = visible
    }

    fun loadMoreItem(searchOffset: Int) {
        this.searchOffset = searchOffset
        updateSearchQuery(searchQuery.value)
    }

    private fun updateTitle(query: String) {
        if (query.isEmpty()) {
            giphyTitle.value = GIF_INITIAL_TITLE
        } else {
            giphyTitle.value = GIF_SEARCH_PREFIX + query + GIF_SEARCH_POSTFIX
        }
    }

    private fun getRandomGifList(count: Int) {
        randomApiJob = CoroutineScope(defaultDispatcher).launch {
            (0 until count)
                .asSequence()
                .asFlow()
                .onEach { delay(API_CALL_DELAY) }
                .flatMapConcat { giphyEngine.getRandomGif() }
                .collect { entity ->
                    val uiModel = GifUiModel(entity.title, entity.url, entity.downsizedUrl)
                    logger.log(java.util.logging.Level.ALL, uiModel.toString())
                    _gifListStateFlow.emit(_gifListStateFlow.value + listOf(uiModel))
                }
        }
    }

    private fun getGifObjectFromQuery(query: String) {
        queryApiJob = CoroutineScope(defaultDispatcher).launch {
            giphyEngine.getGifFromSearchQuery(query, searchOffset)
                .collectLatest { newList ->
                    val newUiModelList =
                        newList.map { GifUiModel(it.title, it.url, it.downsizedUrl) }
                    _gifListStateFlow.emit(gifListStateFlow.value + newUiModelList)
                }
        }
    }

    private fun getAutoCompleteTerms(query: String) {
        if (query.length < 2 || autoTermsList.isNotEmpty()) return

        CoroutineScope(defaultDispatcher).launch {
            giphyEngine.getAutoCompleteTerms(query)
                .collectLatest { autoTermList ->
                    val uiModelList = autoTermList.map { GifAutoCompleteUiModel(it.name) }
                    if (uiModelList.isNotEmpty()) {
                        setAutoCompleteVisibility(true)
                        autoTermsList.addAll(uiModelList)
                    }
                }
        }
    }

    fun cancelAllJob() {
        allApiJob.cancel()
    }

    companion object {
        private const val GIF_INITIAL_TITLE = "Giphy Random GIFs"
        private const val GIF_SEARCH_PREFIX = "Search "
        private const val GIF_SEARCH_POSTFIX = " ..."
        private const val API_CALL_DELAY = 200L

        private const val GIF_COUNT = 30
    }
}
